Tensor parallel distributed strategy without using deepspeed

kalyanjk opened the pull request:
Tensor parallel by extending GaudiLlamaAttention -> TPGaudiLlamaAttention and GaudiLlamaMLP -> TPGaudiLlamaMLP

use parameter --distributed_strategy="tp" to invoke this code path

code design reference: https://github.com/foundation-model-stack/foundation-model-stack/tree/main

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_1121). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

regisss commented on optimum/habana/distributed/serialization.py:
```diff
@@ -0,0 +1,489 @@
+
+import torch
+
+from optimum.habana.distributed.tp import TPModule
```
```suggestion
from .tp import TPModule
```

kalyanjk replied:
Done

regisss commented on optimum/habana/distributed/serialization.py:
```diff
@@ -0,0 +1,489 @@
+    return adapted
+
+
+# `models` imports each model class, causing models and adapters to be registered.
+# down here to avoid circular dependencies.
+# from fms import models
```
Do we need this?

kalyanjk replied:
Removed

regisss commented on optimum/habana/transformers/models/llama/modeling_llama.py:
```diff
@@ -20,10 +20,20 @@
     _gaudi_prepare_4d_causal_attention_mask,
 )
 
+from optimum.habana.distributed.tp import TPModule
```
```suggestion
from ....distributed.tp import TPModule
```

kalyanjk replied:
Done

regisss commented on optimum/habana/transformers/models/llama/modeling_llama.py:
```diff
@@ -20,10 +20,20 @@
 )
 
+from optimum.habana.distributed.tp import TPModule
+from  optimum.habana import distributed
```
```suggestion
from .... import distributed
```

kalyanjk replied:
Done

regisss commented on optimum/habana/transformers/models/llama/modeling_llama.py:
```diff
@@ -20,10 +20,20 @@
 
+from optimum.habana.distributed.tp import TPModule
+from  optimum.habana import distributed
+from optimum.habana.distributed.tensorparallel import (
```
```suggestion
from ....distributed.tensor_parallel import (
```

kalyanjk replied:
Done

regisss commented on optimum/habana/transformers/models/llama/modeling_llama.py:
```diff
@@ -20,10 +20,20 @@
+    reduce_from_tensor_model_parallel_region,
+)
+
+from optimum.habana.distributed.strategy import DistributedStrategy
```
```suggestion
from ....distributed.strategy import DistributedStrategy
```

kalyanjk replied:
Done

regisss commented on optimum/habana/transformers/models/llama/modeling_llama.py:
```diff
@@ -20,10 +20,20 @@
+)
+
+from optimum.habana.distributed.strategy import DistributedStrategy
+from optimum.habana.distributed.strategy import  NoOpStrategy
```
```suggestion
from ....distributed.strategy import NoOpStrategy
```

kalyanjk replied:
Done

regisss commented on optimum/habana/transformers/models/llama/modeling_llama.py:
```diff
@@ -717,10 +855,16 @@ def __init__(self, config: LlamaConfig):
         self.padding_idx = config.pad_token_id
         self.vocab_size = config.vocab_size
 
+        self.distributed_strategy = config.distributed_strategy
```
You need to add `distributed_strategy` to `LlamaConfig` here: https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/models/llama/configuration_llama.py
You can follow what is done for `mlp_bias` and `fused_qkv`.

kalyanjk replied:
Done

regisss replied:
Hmm I don't see this change

kalyanjk replied:
I had removed this change  self.distributed_strategy = config.distributed_strategy
so it thought it was not necessary since i dint see it getting used any where.
I have updated the  LlamaConfig now.

regisss replied:
Yeah, it may not be absolutely necessary but I feel like it's cleaner that way.

regisss commented on optimum/habana/transformers/models/llama/modeling_llama.py:
```diff
@@ -717,10 +855,16 @@ def __init__(self, config: LlamaConfig):
         self.vocab_size = config.vocab_size
 
+        self.distributed_strategy = config.distributed_strategy
+        config.distributed_strategy = None
```
Why is this needed?

kalyanjk replied:
Removed

regisss reviewed (commented):
I left a few comments. Additionally, can you:
- run `make style`?
- add an example command in the README of the text-generation example?
- add a test for it in https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py?
- add a link to the original implementation in all files that are inspired from it?
- check if there are any copyrights to cite?

kalyanjk commented:
> I left a few comments. Additionally, can you:
> 
> * run `make style`?
> * add an example command in the README of the text-generation example?
> * add a test for it in https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py?
> * add a link to the original implementation in all files that are inspired from it?
> * check if there are any copyrights to cite?

Done

kalyanjk commented on examples/text-generation/utils.py:
```diff
@@ -245,6 +245,102 @@ def setup_model(args, model_dtype, model_kwargs, logger):
+
+        def __setstate__(self, state):
+            self.__dict__.update(state)
+            self.group = None  # Restore to default state or reinitialize
```
Done

kalyanjk commented on examples/text-generation/utils.py:
```diff
@@ -245,6 +245,102 @@ def setup_model(args, model_dtype, model_kwargs, logger):
+        import habana_frameworks.torch.distributed.hccl
+        dist.init_process_group(backend='hccl')
+    else:
+        dist.init_process_group()
```
Done

kalyanjk commented on examples/text-generation/utils.py:
```diff
@@ -245,6 +245,102 @@ def setup_model(args, model_dtype, model_kwargs, logger):
+    if args.use_hpu_graphs:
+        from habana_frameworks.torch.hpu import wrap_in_hpu_graph
+
+        if check_habana_frameworks_version("1.13.0") and model.config.model_type == "falcon":
+            model = wrap_in_hpu_graph(model, hash_with_views=False)
+        else:
```
Done

kalyanjk commented on optimum/habana/distributed/tp_wrapping.py:
```diff
@@ -0,0 +1,33 @@
+    TPGaudiLlamaAttention
+)
+
+# this probably belongs somewhere else but can't go in fms.distribtued b/c
+# circular dependency.
```
Done

kalyanjk commented on optimum/habana/distributed/tp.py:
```diff
@@ -0,0 +1,84 @@
+import torch.nn as nn
+from torch.distributed.distributed_c10d import ProcessGroup
+
+from optimum.habana.distributed.tensorparallel import (
```
Done

kalyanjk commented on optimum/habana/distributed/tp_wrapping.py:
```diff
@@ -0,0 +1,33 @@
+from torch import nn
+from torch.distributed.distributed_c10d import ProcessGroup
+
+from optimum.habana.transformers.models.llama.modeling_llama import (
```
Done

kalyanjk commented on optimum/habana/distributed/tensorparallel.py:
```diff
@@ -0,0 +1,103 @@
+# mypy: disable-error-code="method-assign,misc"
```
Done

kalyanjk commented:
@regisss Addressed all your comments. Please review the changes. Thank you!

regisss commented on tests/test_text_generation_example.py:
```diff
@@ -71,6 +71,9 @@
             ("meta-llama/Llama-2-7b-hf", 39.72973199515235),
         ],
+        "distributed_tp": [
+            ("/mnt/weka/data/pytorch//llama2/Llama-2-7b-hf/", 1856.8140409694543),
```
```suggestion
            ("meta-llama/Llama-2-7b-hf", 1856.8140409694543),
```
This is your local path, let's switch to the hub id.

kalyanjk replied:
> This is your local path, let's switch to the hub id.

Done

regisss commented on tests/test_text_generation_example.py:
```diff
@@ -142,11 +152,13 @@ def _test_text_generation(
 
     if torch_compile:
         command += ["--torch_compile"]
+        command += ["--use_flash_attention"]
+        command += ["--flash_attention_recompute"]
```
Let's add these only if `distributed_strategy == "tp"`

kalyanjk replied:
Done

regisss commented on examples/text-generation/README.md:
```diff
@@ -264,6 +264,35 @@ set the following environment variables before running the command: `PT_ENABLE_I
 
 You will also need to add `--torch_compile` in your command.
 
+### Running with Tesor parallel strategy
```
```suggestion
### Running with tensor-parallel strategy
```

kalyanjk replied:
Done

regisss commented on examples/text-generation/README.md:
```diff
@@ -264,6 +264,35 @@ set the following environment variables before running the command: `PT_ENABLE_I
+### Running with Tesor parallel strategy
+#### Attribution
+
+This repository includes code from the [foundation-model-stack](https://github.com/foundation-model-stack/foundation-model-stack) repository, which is licensed under the Apache License 2.0. See the `LICENSE` file for more details.
```
```suggestion
> [!NOTE]  
> This strategy includes code from the [foundation-model-stack](https://github.com/foundation-model-stack/foundation-model-stack) repository, which is licensed under the Apache License 2.0. See the `LICENSE` file for more details.
```

kalyanjk replied:
Done

regisss commented on examples/text-generation/README.md:
```diff
@@ -264,6 +264,35 @@ set the following environment variables before running the command: `PT_ENABLE_I
 You will also need to add `--torch_compile` in your command.
 
+### Running with Tesor parallel strategy
+#### Attribution
```
I think you can remove that line, let's put it in a "box" as suggested below

kalyanjk replied:
I have added in the box, but not sure if this syntax had to be preserved [!WARNING]

regisss commented on examples/text-generation/README.md:
```diff
@@ -264,6 +264,35 @@ set the following environment variables before running the command: `PT_ENABLE_I
+
+This repository includes code from the [foundation-model-stack](https://github.com/foundation-model-stack/foundation-model-stack) repository, which is licensed under the Apache License 2.0. See the `LICENSE` file for more details.
+
+torch.compile with tensor parallel strategy is an experimental feature. It has not been validated for all models. To enable
```
```suggestion
> [!WARNING]  
> torch.compile with tensor parallel strategy is an experimental feature. It has not been validated for all models.

To enable...
```

kalyanjk replied:
Done, Added Note and Warning in box

regisss reviewed (commented):
Can you update your main branch and merge it into this PR? To have the whole CI working again.

libinta commented:
@kalyanjk can you update based on review comments?

kalyanjk commented:
> @kalyanjk can you update based on review comments?

@libinta, I have addressed all.  Anything i have missed?

regisss reviewed (commented):
I left a few more comments to address.
Also, the test fails on my instance with this error:
```
Traceback (most recent call last):                                                                                                           
  File "/root/workspace/fork/examples/text-generation/run_generation.py", line 674, in <module>                                              
    main()                                                                                                                                   
  File "/root/workspace/fork/examples/text-generation/run_generation.py", line 317, in main                                                  
    model, assistant_model, tokenizer, generation_config = initialize_model(args, logger)                                                    
  File "/root/workspace/fork/examples/text-generation/utils.py", line 592, in initialize_model                                               
    else setup_distributed_model_tp(args, model_dtype, model_kwargs, logger)                                                                 
  File "/root/workspace/fork/examples/text-generation/utils.py", line 281, in setup_distributed_model_tp                                     
    lazy_sd = serialization.load_state_dict(                                                                                                 
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/distributed/serialization.py", line 191, in load_state_dict                   
    assert len(checkpoints) > 0, f"Can't find the requested checkpoint data at {model_path}"                                                 
AssertionError: Can't find the requested checkpoint data at meta-llama/Llama-2-7b-hf
```
Any idea about what's going on? It seems like a serialization issue. Or is it because it requires Synapse 1.17? I'm running 1.16.

regisss commented:
Please run `make style` too

kalyanjk commented on examples/text-generation/README.md:
````diff
@@ -264,6 +264,38 @@ set the following environment variables before running the command: `PT_ENABLE_I
+
+Here is an example:
+```bash
+python ../gaudi_spawn.py  --world_size 8 run_generation.py \
````
updated

kalyanjk commented on examples/text-generation/README.md:
````diff
@@ -264,6 +264,38 @@ set the following environment variables before running the command: `PT_ENABLE_I
+WARNING: torch.compile with tensor parallel strategy is an experimental feature. It has not been validated for all models.
+```
+To enable torch.compile with tensor parallel strategy, please set the following environment variables before running the
+command: `PT_ENABLE_INT64_SUPPORT=1` and `PT_HPU_LAZY_MODE=0`. This will enable tensor parallel strategy without deepspeed.
````
updated with the suggested format

kalyanjk commented on examples/text-generation/README.md:
````diff
@@ -264,6 +264,38 @@ set the following environment variables before running the command: `PT_ENABLE_I
+
+```bash
+NOTE: This strategy includes code from the [foundation-model-stack](https://github.com/foundation-model-stack/foundation-model-stack) repository, which is licensed under the Apache License 2.0. See the `LICENSE` file for more details.
+```
````
updated with the suggested format

kalyanjk commented on examples/text-generation/README.md:
```diff
@@ -264,6 +264,38 @@ set the following environment variables before running the command: `PT_ENABLE_I
 
 You will also need to add `--torch_compile` in your command.
 
+### Running with tesor-parallel strategy
```
Done

kalyanjk commented:
> I left a few more comments to address. Also, the test fails on my instance with this error:
> 
> ```
> Traceback (most recent call last):                                                                                                           
>   File "/root/workspace/fork/examples/text-generation/run_generation.py", line 674, in <module>                                              
>     main()                                                                                                                                   
>   File "/root/workspace/fork/examples/text-generation/run_generation.py", line 317, in main                                                  
>     model, assistant_model, tokenizer, generation_config = initialize_model(args, logger)                                                    
>   File "/root/workspace/fork/examples/text-generation/utils.py", line 592, in initialize_model                                               
>     else setup_distributed_model_tp(args, model_dtype, model_kwargs, logger)                                                                 
>   File "/root/workspace/fork/examples/text-generation/utils.py", line 281, in setup_distributed_model_tp                                     
>     lazy_sd = serialization.load_state_dict(                                                                                                 
>   File "/usr/local/lib/python3.10/dist-packages/optimum/habana/distributed/serialization.py", line 191, in load_state_dict                   
>     assert len(checkpoints) > 0, f"Can't find the requested checkpoint data at {model_path}"                                                 
> AssertionError: Can't find the requested checkpoint data at meta-llama/Llama-2-7b-hf
> ```
> 
> Any idea about what's going on? It seems like a serialization issue. Or is it because it requires Synapse 1.17? I'm running 1.16.

Can you provide with absolute path for meta-llama/Llama-2-7b-hf.  All my testing is on 1.17, I will verify on 1.16 and update.

kalyanjk commented:
> > I left a few more comments to address. Also, the test fails on my instance with this error:
> > ```
> > Traceback (most recent call last):                                                                                                           
> >   File "/root/workspace/fork/examples/text-generation/run_generation.py", line 674, in <module>                                              
> >     main()                                                                                                                                   
> >   File "/root/workspace/fork/examples/text-generation/run_generation.py", line 317, in main                                                  
> >     model, assistant_model, tokenizer, generation_config = initialize_model(args, logger)                                                    
> >   File "/root/workspace/fork/examples/text-generation/utils.py", line 592, in initialize_model                                               
> >     else setup_distributed_model_tp(args, model_dtype, model_kwargs, logger)                                                                 
> >   File "/root/workspace/fork/examples/text-generation/utils.py", line 281, in setup_distributed_model_tp                                     
> >     lazy_sd = serialization.load_state_dict(                                                                                                 
> >   File "/usr/local/lib/python3.10/dist-packages/optimum/habana/distributed/serialization.py", line 191, in load_state_dict                   
> >     assert len(checkpoints) > 0, f"Can't find the requested checkpoint data at {model_path}"                                                 
> > AssertionError: Can't find the requested checkpoint data at meta-llama/Llama-2-7b-hf
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Any idea about what's going on? It seems like a serialization issue. Or is it because it requires Synapse 1.17? I'm running 1.16.
> 
> Can you provide with absolute path for meta-llama/Llama-2-7b-hf. All my testing is on 1.17, I will verify on 1.16 and update.

@regisss successfully verified the sanity test for the 1.16 release using both the 7b and 70b models. Everything is working fine.

regisss commented:
There is no absolute path, this is the hub model id and I really think this use case should work as not everybody has the models stored locally. If the absolute path to the model is needed, there should be some code to find the model in the Transformers cache. You can get the default path to cache with:
```
from huggingface_hub.constants import HF_HUB_CACHE
```
More information about the structure of the cache here: https://huggingface.co/docs/huggingface_hub/v0.24.2/en/guides/manage-cache#understand-caching

Also, I see I forgot to mention it, can you replace the arg `distributed_strategy` by `parallelism_strategy` or something similar everywhere please? We already have `distribution_strategy` defined [here](https://github.com/huggingface/optimum-habana/blob/a4851cde758fe8e50044f43d1104cbd2e77b1ff6/optimum/habana/transformers/training_args.py#L169) and it would add a lot of confusion to have both `distributed_strategy` and `distribution_strategy`. Sorry for noticing this now, I thought I had commented on it before.

kalyanjk commented:
> There is no absolute path, this is the hub model id and I really think this use case should work as not everybody has the models stored locally. If the absolute path to the model is needed, there should be some code to find the model in the Transformers cache. You can get the default path to cache with:
> 
> ```
> from huggingface_hub.constants import HF_HUB_CACHE
> ```
> 
> More information about the structure of the cache here: https://huggingface.co/docs/huggingface_hub/v0.24.2/en/guides/manage-cache#understand-caching
> 
> Also, I see I forgot to mention it, can you replace the arg `distributed_strategy` by `parallelism_strategy` or something similar everywhere please? We already have `distribution_strategy` defined [here](https://github.com/huggingface/optimum-habana/blob/a4851cde758fe8e50044f43d1104cbd2e77b1ff6/optimum/habana/transformers/training_args.py#L169) and it would add a lot of confusion to have both `distributed_strategy` and `distribution_strategy`. Sorry for noticing this now, I thought I had commented on it before.

Updated : renamed the distributed_strategy to parallel_strategy.  
In process : Is there a way i can test the relative path for model_name - > meta-llama/Llama-2-7b-hf.

kalyanjk commented:
> > There is no absolute path, this is the hub model id and I really think this use case should work as not everybody has the models stored locally. If the absolute path to the model is needed, there should be some code to find the model in the Transformers cache. You can get the default path to cache with:
> > ```
> > from huggingface_hub.constants import HF_HUB_CACHE
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > More information about the structure of the cache here: https://huggingface.co/docs/huggingface_hub/v0.24.2/en/guides/manage-cache#understand-caching
> > Also, I see I forgot to mention it, can you replace the arg `distributed_strategy` by `parallelism_strategy` or something similar everywhere please? We already have `distribution_strategy` defined [here](https://github.com/huggingface/optimum-habana/blob/a4851cde758fe8e50044f43d1104cbd2e77b1ff6/optimum/habana/transformers/training_args.py#L169) and it would add a lot of confusion to have both `distributed_strategy` and `distribution_strategy`. Sorry for noticing this now, I thought I had commented on it before.
> 
> Updated : renamed the distributed_strategy to parallel_strategy. In process : Is there a way i can test the relative path for model_name - > meta-llama/Llama-2-7b-hf.

Updated cache_dir setting for  parallel_strategy = tp

@regisss  can you please verify if you are able to load the data now

regisss commented on tests/test_text_generation_example.py:
```diff
@@ -71,6 +71,9 @@
             ("meta-llama/Llama-2-7b-hf", 39.72973199515235),
         ],
+        "distributed_tp": [
+            ("meta-llama/Llama-2-7b-hf", 1856.8140409694543),
```
I get 1345.2369318328463 on my Gaudi2 instance with 1.16. Could you check if you get the same? I'm assuming the gap comes from 1.17, but I'm really not sure about it.

kalyanjk replied:
@regiss I observed a performance number of 1361.2050636537415 on the 1.16 release, which is close what you have observed. I have updated the expected performance number accordingly.

regisss reviewed (commented):
Thanks for the changes, it looks good to me!
One last thing, as written in the comment below, the test fails on my instance because the throughput I get is too low. Maybe due to a different version of Synapse?
